docs/quarantine.md

```diff
@@ -0,0 +1,76 @@
+# Test Quarantine
+
+The execution of tests on CI should be as deterministic as possible, besides
+potential issues coming from infrastructure or external dependencies. When a
+test sometimes fails and sometimes passes, apparently randomly, without any
+relationship with the changes in a PR, it has very negative implications:
+* The results of the whole suite are less valuable, we can no longer trust
+the information given by the suite execution. Changes are added to the codebase
+without actually knowing if they are going to break something.
+* As a consequence, we can't be completely sure about the state of the product
+at release time, it can be ok or not.
+* The development process is affected negatively. Developers need to wait
+until a lucky execution aligns all the required tests in green, which can take
+more than a week in some cases.
+
+In summary, the CI system stops helping the product evolution and instead becomes
+an obstacle.
+
+We have introduced a quarantine methodology to put apart the tests that don't
+behave deterministically until they are fixed, so that we can keep the rest of
+the suite as healthy as possible. You can read more about test quarantine
+methodology in [1] and [2], and about an actual implementation in [3].
+
+## Procedure
+
+In order to remove as much as possible the influence of changes in PRs to
+determine the stability of the suite, we will take into account only results
+from the periodics that run e2e tests from master, these jobs can be checked
+[on testgrid].
+
+In the following definitions we will consider test failures only when reported on
+testgrid in jobs where less than 5 tests failed, so that we don't take into
+account systemic failures caused for instance by an infrastructure problem.
+
+### Putting tests in quarantine
+
+A test must be put in quarantine when any of these conditions is met:
+* It has failed 2 times or more in the last two weeks.
+* It has failed 5 times or more in the last 3 days.
```

Comment:
@fgimenez I've been thinking about these thresholds a little more. Depending on the sample size of test results, they may or may not be accurate for us.

For example, 2 failures over two weeks for a sample size of 200 test runs is a 1% failure rate... that's probably not worth us quarantining

However, 2 failures over two weeks with a sample size of 30 test runs is a something like a 6% failure rate, which to me crosses a threshold of needing attention.

What if we consider a failure percentage as our threshold rather than number of times failed? For me, I'd consider a 95% pass rate (5% failure rate) as a reasonable threshold as long as we have a sample size over 100.

Reply:
Something like this would really be preferrable for me.

A few more considerations:
 * I would also suggest to not use jobs from PRs which were not merged. This means that we can only take what flakefinder already takes (presumbit runs on then finally merged code) + periodic jobs. I think we should look into creating a reasonable mass of runs with these.
  * I would also consider a grace period for tests with a certain percentage (of course not if they fail e.g. 20% of the time). There should be the chance to fix the tests within e.g. one or two days before they are moved to quarntine, once one notices that tests need attention. Otherwise one would have to wait at least a week to unquarantine them via automated processes, which leave blind spots.

Reply:
> For me, I'd consider a 95% pass rate (5% failure rate) as a reasonable threshold as long as we have a sample size over 100.

totally agree, will update the doc accordingly.

> * I would also suggest to not use jobs from PRs which were not merged. This means that we can only take what flakefinder already takes (presumbit runs on then finally merged code) + periodic jobs.

makes sense to me, this removes a lot of noise.

> * I would also consider a grace period for tests with a certain percentage

This is very interesting, similar to what we have already in place for release blockers. 2 days of grace period since the quarantine PR is proposed sounds very good to me, I'll add that to the doc.

Reply:
> Otherwise one would have to wait at least a week to unquarantine them via automated processes, which leave blind spots.

In any case we need to keep in mind that the test is still giving feedback, we keep executing it in the quarantined lanes (for now only periodics).

Reply:
> In any case we need to keep in mind that the test is still giving feedback, we keep executing it in the quarantined lanes (for now only periodics).

For that one has to know that a relevant test for a code change exists and that it was moved to quarantine. Further we expect then that people start looking into lanes where things are almost "supposed" to fail, while at the same time saying we move flaky tests out as fast as possible to avoid that people find a flaky test not relevant.

I find it great that we run them, and experienced maintainers can maybe still make something out of them, but I would not bring it up here as an argument that we still get feedback.

Reply:
Ok pushed changes to address your comments @davidvossel @rmohr PTAL.

For the grace period I've included a number of successful executions for taking a fix into account, let me know wdyt.

Reply:
> Ok pushed changes to address your comments @davidvossel @rmohr PTAL.

yep, new wording seems reasonable. I'd consider all these values as a starting point for us. Lets remain remain flexible to adjust these if we find our percentages or time frames don't end up serving us well.

Reply:
>>  For me, I'd consider a 95% pass rate (5% failure rate) as a reasonable threshold as long as we have a sample size over 100.
> totally agree, will update the doc accordingly.

@davidvossel , @fgimenez : to me, a 95% pass rate can be too low for a single test. With 16 tests currently quarantined, it puts us at 95%**16=44% pass rate for the entire suite. e2e suite failure must be rare, so that when it happens everyone is startled and goes to check why did it happen. After some 5 weeks below 50%, we are on 80% today according to [grafana](https://grafana-kubevirt-prow.apps.ovirt.org/d/Cz6zgEEGz/e2e-jobs-overview?orgId=1&refresh=2h&from=now-14d&to=now&var-job_name=periodic-kubevirt-e2e-k8s-.*(prev%7Clatest)). I'd consider a 95% pass rate as the bare minimum for the entire per-pull suites, which means a pass rate higher than 99.5% for an individual test.

Reply:
Effectively, when looking at the whole quarantine flow:
 * test get into quarantine
 * someone gets assigned to fix the test
 * The person needs X days to find the issue
 * The test needs to succeed for one week from there on
 
how do we prevent having coverage holes for weeks on PRs?

Reply:
To me, the interesting parameter is not an individual test pass rate but the total pass rate of the suites we run per PR. When tests fail on a PR, the PR author must be 95% sure that it their fault. Otherwise they blindly /retest and waste CI resources. 

> how do we prevent having coverage holes for weeks on PRs?

Suggestion: during the X+7 days of quarantine, the person assigned to the test must keep testing it on master. They must raise the alarm if suddenly their test fails consistently, and have the faulty PR reverted.

Reply:
> Effectively 99.5% means a lot of tests are candidates for quarantine.

In fact only the tests that are actually preventing PRs from being merged because of non-determinism, if a test behaves properly then it won't be proposed even with this lower failure percentage.

> how do we prevent having coverage holes for weeks on PRs?

ATM we have the periodics that run the whole suite (including quarantined tests) 3 times per day to catch any error that went unnoticed in PRs, as soon as resources are in place we can run quarantine lanes also on PRs. In any case I wonder what coverage gives us a flaky test, we can't trust the information it provides. In my opinion proper test coverage comes from deterministic tests.

Reply:
> > how do we prevent having coverage holes for weeks on PRs?
> 
> Suggestion: during the X+7 days of quarantine, the person assigned to the test must keep testing it on master. They must raise the alarm if suddenly their test fails consistently, and have the faulty PR reverted.

I think we should have this as part of the stability check for getting tests out of quarantine. After the fix has landed, the assigned person needs to check daily if the quarantine test executions from master (in periodics) have any failure. When periodics are integrated into flakefinder this will be easier to check.

Reply:
> I'd consider a 95% pass rate as the bare minimum for the entire per-pull suites, which means a pass rate higher than 99.5% for an individual test.

Agree and updated accordingly, PTAL.

Reply:
> For that one has to know that a relevant test for a code change exists and that it was moved to quarantine. Further we expect then that people start looking into lanes where things are almost "supposed" to fail, while at the same time saying we move flaky tests out as fast as possible to avoid that people find a flaky test not relevant.
> 
> I find it great that we run them, and experienced maintainers can maybe still make something out of them, but I would not bring it up here as an argument that we still get feedback.

As I see it this is where test owners come into play, they need to take care of bringing the test out of quarantine and should be aware of any failures during the quarantine period. 

I agree that keeping track of all the executions on all the PRs (to a lesser extent also all the periodic executions) can be hard, but we could help here with automation, maybe drop a comment on a PR that had a failing quarantine test tagging the owner, or sending an email to the owner after a quarantined test failed on a periodic, wdyt?